### PR TITLE
VS-146: Support Fully Qualified Names for Data Models

### PR DESCRIPTION
**Pull request by Ravi-Raghavan**
Support Fully Qualified Names for Data Models

### CODE REVIEW

File: tests/MongoDB.Analyzer.Tests.Common.TestCases/Builders/BuildersQualifiedNames.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+                Builders<Person>.Filter.Eq(p => p.SiblingsCount, MongoDB.Analyzer.Tests.Common.DataModel.StaticHolder.ReadonlyShort) |
+                Builders<Person>.Filter.Eq(p => p.SiblingsCount, MongoDB.Analyzer.Tests.Common.DataModel.StaticHolder.ReadonlyInt) |
+                Builders<Person>.Filter.Eq(p => p.TicksSinceBirth, MongoDB.Analyzer.Tests.Common.DataModel.StaticHolder.ReadonlyLong) |
+                Builders<Person>.Filter.Eq(p => p.Name, MongoDB.Analyzer.Tests.Common.DataModel.StaticHolder.ReadonlyString) |
+                Builders<Person>.Filter.Eq(p => p.Name, MongoDB.Analyzer.Tests.Common.DataModel.StaticHolder.ReadonlyString) |
+                Builders<Person>.Filter.Eq(p => p.Name, MongoDB.Analyzer.Tests.Common.DataModel.StaticHolder.ReadonlyString2) |
+                Builders<Person>.Filter.Eq(p => p.Name, MongoDB.Analyzer.Tests.Common.DataModel.StaticHolder.ReadonlyString) |
+                Builders<Person>.Filter.Eq(p => p.Vehicle.VehicleType.Type, MongoDB.Analyzer.Tests.Common.DataModel.StaticHolder.ReadonlyEnum) |
```
Comment (BorisDog) [inline tests/MongoDB.Analyzer.Tests.Common.TestCases/Builders/BuildersQualifiedNames.cs:68]: Same naming pattern, no need to test all types.
Also consider combing Qualified_inside_method and Qualified_inside_namespace into single test with different naming nesting levels. Given `A.B...X.Y.Class.Property` is a full name, let's test:
```
A.B...X.Y.Class.Propert
B...X.Y.Class.Propert
X.Y.Class.Propert
Y.Class.Propert
```
Same for `Builders<A.B...X.Y.Class>`.

Also  test cases like:

```
using user = MongoDB.Analyzer.Tests.Common.DataModel.User;
using driver = MongoDB.Driver;
_ = driver.Builders<user>.Filter.Eq(user => user.Age, 22);
_ = driver.Builders<MongoDB.Analyzer.Tests.Common.DataModel.User>.Filter.Eq(user => user.Age, 22);
```

**Review by BorisDog — CHANGES_REQUESTED** [review]

File: tests/MongoDB.Analyzer.Tests.Common.TestCases/Builders/BuildersQualifiedNames.cs (hunk 1 above)
Comment (Ravi-Raghavan) [reply tests/MongoDB.Analyzer.Tests.Common.TestCases/Builders/BuildersQualifiedNames.cs]: Done

File: src/MongoDB.Analyzer/Core/Builders/BuilderExpressionProcessor.cs [hunk 2]
```diff
@@ -143,11 +143,32 @@ private static IEnumerable<SyntaxNode> GetBuildersDefinitionNodes(SemanticModel
             {
                 nodesProcessed.Add(node);
 
-                // Skip MongoDB.Driver.Builders<T> nodes
+                var isBuildersContainer = false;
+                var containsAlias = false;
+
+                //Check if Member Expression is Builders Container
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Builders/BuilderExpressionProcessor.cs:149]: Please add space after `//` in all comments

File: src/MongoDB.Analyzer/Core/Builders/BuilderExpressionProcessor.cs [hunk 3]
```diff
@@ -143,11 +143,32 @@ private static IEnumerable<SyntaxNode> GetBuildersDefinitionNodes(SemanticModel
...
+                    if (semanticModel.GetAliasInfo(syntaxNode) != null)
+                    {
+                        containsAlias = true;
+                        break;
+                    }
+
+                    syntaxNode = (syntaxNode as MemberAccessExpressionSyntax)?.Expression;
+                }
+
+                //Skip Nodes that are Builder Containers and that don't contain alias
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Builders/BuilderExpressionProcessor.cs:170]: Add two examples in the comment:
// For example MongoDB.Driver.Builders<T>, driverAlias.Builders<T>

File: src/MongoDB.Analyzer/Core/Builders/BuilderExpressionProcessor.cs [hunk 4]
```diff
@@ -143,11 +143,32 @@ private static IEnumerable<SyntaxNode> GetBuildersDefinitionNodes(SemanticModel
...
-                    namedTypeSymbol.IsBuildersContainer() &&
-                    semanticModel.GetAliasInfo(memberAccessExpressionSyntax.Expression) == null)
+                    namedTypeSymbol.IsBuildersContainer())
+                {
+                    isBuildersContainer = true;
+                }
+
+                //Check if Node contains alias
+                var syntaxNode = (node as MemberAccessExpressionSyntax)?.Expression;
+                while (syntaxNode != null)
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Builders/BuilderExpressionProcessor.cs:159]: SyntaxNodeExtensions.NestedInvocations won't work here?

File: src/MongoDB.Analyzer/Core/Builders/BuilderExpressionProcessor.cs (hunk 2 above)
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Builders/BuilderExpressionProcessor.cs:149]: Done

File: src/MongoDB.Analyzer/Core/Builders/BuilderExpressionProcessor.cs (hunk 3 above)
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Builders/BuilderExpressionProcessor.cs:170]: Done

**Review by BorisDog — APPROVED** [review]
LGTM! Good work!

**Pull request merged**